lib/improver/nbhood.py

```diff
@@ -195,6 +198,75 @@ def __repr__(self):
+        Method to calculate the average value in a square neighbourhood using
+        the 4-point algorithm to find the total sum over the neighbourhood.
+
+        The output from the cumulate_array method can be used to
+        calculate the sum over a neighbourhood of size cells_x*cells_y. This
+        sum is then divided by the area of the neighbourhood to calculate the
+        mean value in the neighbourhood.
+
+        Parameters
+        ----------
+        cube : iris.cube.Cube
+            Cube to which neighbourhood processing is being applied. Must
+            be passed through cumulate_array method first.
+        cells_x, cells_y : integer
+            The width of the neighbourhood in grid cells, in the x and y
+            directions.
+
+        Returns
+        -------
+        cube : iris.cube.Cube
+            Cube to which square neighbourhood has been applied.
+        """
+        def _sum_and_area_of_neighbourhood(cube, cells_x, cells_y, i, j):
+            """
+            Function to calculate the total sum and area of the neighbourhood
+            surrounding a single grid point.
+            """
+            x_min = i-cells_x-1
+            x_max = min(cube.shape[1]-1, i+cells_x)
+            y_min = j-cells_y-1
+            y_max = min(cube.shape[0]-1, j+cells_y)
+            summed_array = cube.data
+            if x_min < 0 and y_min < 0:
+                total = summed_array[y_max, x_max]
+            elif x_min < 0 and y_min >= 0:
+                total = (summed_array[y_max, x_max] -
+                         summed_array[y_min, x_max])
+            elif x_min >= 0 and y_min < 0:
+                total = (summed_array[y_max, x_max] -
+                         summed_array[y_max, x_min])
+            else:
+                total = (summed_array[y_max, x_max] -
+                         summed_array[y_min, x_max] -
+                         summed_array[y_max, x_min] +
+                         summed_array[y_min, x_min])
+            x_min = max(-1, x_min)
+            y_min = max(-1, y_min)
+            area = (y_max - y_min) * (x_max - x_min)
+            return total, area
+
+        yname = cube.coord(axis="y").name()
+        xname = cube.coord(axis="x").name()
+        cubelist = iris.cube.CubeList([])
+        for slice_2d in cube.slices([yname, xname]):
+            sum_over_neighbourhood = np.zeros(slice_2d.shape)
+            neighbourhood_area = np.zeros(slice_2d.shape)
+            # Calculate total sum and area of neighbourhood at each grid point
+            for i in range(slice_2d.shape[1]):
+                for j in range(slice_2d.shape[0]):
+                    sum_over_neighbourhood[j, i], neighbourhood_area[j, i] = (
+                        _sum_and_area_of_neighbourhood(
+                            slice_2d, cells_x, cells_y, i, j))
+            slice_2d.data = sum_over_neighbourhood/neighbourhood_area
+            cubelist.append(slice_2d)
+        return cubelist.merge_cube()
```

Comment:
Speeding this up.

To vectorise this you can make the problem linear (flatten the array), work out the 4 displacements along the flattened array that get the relevant corner points (these will be a constant displacement for a given cell_x/y and grid size), make 4 copies of the array and np.roll these along so that all the indices align. Then it is a simple matter of adding and subtracting the 4 arrays as per the bit above. Hey presto it's all vectorised. You take more memory but a lot let processing time (it is absolutely prohibitively slow at the moment).

I've added an example to the ticket.